File: concepts/10_time_locale_and_runtime/03_locale_numbers_dates_and_collation/lua/test_02_availability_and_state_boundaries.lua
-- Common question: how is locale availability detected and process-global state restored?
-- Inputs: current locale, guaranteed C locale, an unavailable name, category changes, and restoration.
-- Observations: nil capability result, category scope, process-global mutation, and explicit restore.
-- polyglot-family: time_locale_and_runtime
-- polyglot-concept: locale_numbers_dates_and_collation
-- polyglot-related: languages/lua/standard_library/11_io_os_and_processes/test_086_time_date_and_differences.lua

local t = require("support.assertions")
local original = assert(os.setlocale(nil, "all"))
local numeric_original = assert(os.setlocale(nil, "numeric"))

t.truth(os.setlocale("C", "numeric"))
t.matches(os.setlocale(nil, "numeric"), "^C")
t.equal(os.setlocale("polyglot_LOCALE_that_does_not_exist", "numeric"), nil)
t.truth(os.setlocale(numeric_original, "numeric"))
t.equal(os.setlocale(nil, "numeric"), numeric_original)
t.equal(os.setlocale(nil, "all"), original)

-- Locale 名称与可用集合由 OS 决定；runner 还用独立进程阻断跨测试泄漏。
t.done()

File: concepts/10_time_locale_and_runtime/03_locale_numbers_dates_and_collation/r/test_02_availability_and_state_boundaries.R
# polyglot-family: time_locale_and_runtime
# polyglot-concept: locale_numbers_dates_and_collation
# polyglot-related: languages/r/standard_library/13_io_text_time_and_numeric_workflows/
# polyglot-related+: test_099_csv_encodings_locale_and_regular_expressions.R
#
# 共同问题：请求的 locale 不可用怎样表达，修改范围如何恢复。
# 对照观察：`Sys.setlocale` 对不可用 locale 返回空字符串并警告；locale 是进程级状态。

local({
    original <- Sys.getlocale("LC_TIME")
    on.exit(suppressWarnings(Sys.setlocale("LC_TIME", original)), add = TRUE)
    unavailable <- suppressWarnings(Sys.setlocale("LC_TIME", "polyglot_INVALID_LOCALE"))

    stopifnot(
        identical(unavailable, ""),
        identical(Sys.getlocale("LC_TIME"), original)
    )
})
